### PR TITLE
Short

### LINKED ISSUE #202: List of variable types
List jou's types and some suggestions...
| types | bits | signed | unsigned | more | status |
|-------| ---- | ------- | ----------- | ----- |  ------ |
| bytes | 8 | hope add  | yes | / | added |
| short | 16 | yes | no | maybe not add |  notyet |
| int | 32 | yes | hope add | / |  added |
| long | 64 | yes | hape add | / |  added |
| float | 32 | yes | no | unsigned can't support |  added |
| double | 64 | yes | no | unsigned can't support | added |

### PR DESCRIPTION
**Pull request by littlewhitecloud**
fixed #202 (add short type)
Maybe...?

### CODE REVIEW

**Comment by littlewhitecloud** [comment]
ready to work on the self-hosted compiler

File: src/build_cfg.c [hunk 1]
```diff
@@ -816,7 +816,7 @@ static void build_statement(struct State *st, const AstStatement *stmt)
         build_assert(st, &stmt->data.expression);
         break;
 
-    case AST_STMT_PASS: // Do nothing here
+    case AST_STMT_PASS: // It doesn't require any expression, but if you don't give it a expression, it will raise error.
```
Comment (littlewhitecloud) [inline src/build_cfg.c:819]:
```suggestion
    case AST_STMT_PASS:
```

File: src/print.c [hunk 2]
```diff
@@ -58,6 +58,9 @@ static void print_constant(const Constant *c)
 void print_token(const Token *token)
 {
     switch(token->type) {
+    case TOKEN_SHORT:
+        printf("shorter %hd \n", (short)token->data.short_value);
```
Comment (littlewhitecloud) [inline src/print.c:62]:
```suggestion
        printf("shorter %hd\n", (short)token->data.short_value);
```

**Review by littlewhitecloud — COMMENTED** [review]
cleanup

File: self_hosted/token.jou [hunk 3]
```diff
@@ -40,6 +42,8 @@ class Token:
             if isprint(self->byte_value) != 0:
                 printf(" '%c'", self->byte_value)
             printf("\n")
+        elif self->kind == TokenKind::Short:
+            printf("shorter %hd\n", self->short_value)
```
Comment (Akuli) [inline self_hosted/token.jou:46]:
```suggestion
            printf("short %hd\n", self->short_value)
```

File: self_hosted/token.jou [hunk 4]
```diff
@@ -99,7 +103,9 @@ class Token:
 
     def fail_expected_got(self, what_was_expected_instead: byte*) -> void:
         got: byte[100]
-        if self->kind == TokenKind::Int:
+        if self->kind == TokenKind::Short:
+            strcpy(got, "a shorter")
```
Comment (Akuli) [inline self_hosted/token.jou:107]:
```suggestion
            strcpy(got, "a short")
```

File: src/print.c [hunk 5]
```diff
@@ -58,6 +58,9 @@ static void print_constant(const Constant *c)
 void print_token(const Token *token)
 {
     switch(token->type) {
+    case TOKEN_SHORT:
+        printf("shorter %hd\n", (short)token->data.short_value);
```
Comment (Akuli) [inline src/print.c:62]:
```suggestion
        printf("short %hd\n", (short)token->data.short_value);
```

File: src/types.c [hunk 6]
```diff
@@ -85,6 +86,7 @@ void init_types(void)
     }
 
     strcpy(global_state.integers[8][false].type.name, "byte");
+    strcpy(global_state.integers[16][false].type.name, "short");
```
Comment (Akuli) [inline src/types.c:89]: The signed type should be named `short`, not the unsigned type:

```suggestion
    strcpy(global_state.integers[16][true].type.name, "short");
```

File: tests/should_succeed/add_sub_mul_div_mod.jou [hunk 7]
```diff
@@ -38,9 +38,11 @@ def main() -> int:
     printf("%d\n", 'a'+'a'+'a') # Output: 35
 
     # Integer overflow. Not UB in Jou.
+    printf("%hd\n", 20000 + 20000) # Output: -25536
```
Comment (Akuli) [inline tests/should_succeed/add_sub_mul_div_mod.jou:41]: This does the addition with `int`s, and then passes an `int` to the `printf()` function when it is expecting a float. Let's instead add `short`s:

```suggestion
    printf("%hd\n", 20000s + 20000s) # Output: -25536
```

**Review by Akuli — COMMENTED** [review]
nice :)

File: src/parse.c [hunk 8]
```diff
@@ -14,6 +14,7 @@ static noreturn void fail_with_parse_error(const Token *token, const char *what_
 {
     char got[200];
     switch(token->type) {
+        case TOKEN_SHORT: strcpy(got, "a shorter"); break;
```
Comment (Akuli) [inline src/parse.c:17]:
```suggestion
        case TOKEN_SHORT: strcpy(got, "a short"); break;
```

File: tests/should_succeed/add_sub_mul_div_mod.jou (hunk 7 above)
Comment (littlewhitecloud) [reply tests/should_succeed/add_sub_mul_div_mod.jou:41]: - **Short literals** are just like in literals, except that they have an extra `S` at the end (e.g. `114S`),
    The `S` must be uppercase.
```suggestion
    printf("%hd\n", 20000S + 20000S) # Output: -25536
```

**Comment by Akuli** [comment]
Thanks!

**Pull request merged**